lib/signs_ui/alerts/state.ex

```diff
@@ -4,79 +4,83 @@ defmodule SignsUi.Alerts.State do
-
-  @type route_id :: String.t()
-  @type alert_map :: %{Alert.id() => Alert.t()}
-  @type route_alerts_map :: %{route_id() => alert_map()}
-
-  @type state :: %__MODULE__{
-          alerts: route_alerts_map()
-        }
+  @type t :: %{Alert.id() => Alert.multi_route()}
 
+  @spec start_link(keyword) :: :ignore | {:error, any} | {:ok, pid}
   def start_link(opts \\ []) do
     {start_opts, init_opts} = Keyword.split(opts, [:name])
-    start_opts = Keyword.put_new(start_opts, :name, __MODULE__)
-    GenServer.start_link(__MODULE__, init_opts, start_opts)
+    GenStage.start_link(__MODULE__, init_opts, start_opts)
   end
 
+  @impl GenStage
   def init(opts) do
-    interval_ms = Keyword.get(opts, :interval_ms, 15_000)
-    :timer.send_interval(interval_ms, :twiddle_state)
-    {:ok, %__MODULE__{}}
+    {:consumer, %{}, opts}
   end
 
-  @spec active_alert_ids(GenServer.server()) :: MapSet.t(Alert.id())
+  @impl GenStage
+  def handle_info(msg, state) do
+    Logger.warn("#{__MODULE__} unknown_message #{inspect(msg)}")
+    {:noreply, [], state}
+  end
+
+  @spec active_alert_ids(GenStage.stage()) :: MapSet.t(Alert.id())
   def active_alert_ids(pid \\ __MODULE__) do
-    GenServer.call(pid, :active_alert_ids)
+    GenStage.call(pid, :active_alert_ids)
   end
 
+  @impl GenStage
   def handle_call(:active_alert_ids, _from, state) do
-    alert_ids =
-      for {_route_id, route_alerts} <- state.alerts,
-          {alert_id, _alert_data} <- route_alerts,
-          do: alert_id
+    alert_ids = state |> Map.keys() |> MapSet.new()
 
-    {:reply, MapSet.new(alert_ids), state}
+    {:reply, alert_ids, [], state}
   end
 
-  def handle_info(:twiddle_state, state) do
-    new_alert_state =
-      case :rand.uniform(5) do
-        1 ->
-          state.alerts |> Enum.to_list() |> Enum.shuffle() |> List.delete_at(0) |> Map.new()
-
-        _ ->
-          route_id = Enum.random(["Red", "Orange", "Blue", "Green-B"])
-          alert_id = 10_000 |> :rand.uniform() |> to_string()
-          service_effect = "Alert #{alert_id} service text"
-
-          Map.merge(
-            state.alerts,
-            %{
-              route_id => %{
-                alert_id => %Alert{
-                  id: alert_id,
-                  service_effect: service_effect,
-                  created_at: DateTime.now!("America/New_York")
-                }
-              }
-            }
-          )
-      end
-
-    SignsUiWeb.Endpoint.broadcast!("signs:all", "new_alert_state", new_alert_state)
-    Logger.info("new_alert_state: #{inspect(new_alert_state)}")
-
-    {:noreply, put_in(state.alerts, new_alert_state)}
+  @impl GenStage
+  @spec handle_events([Event.t()], GenStage.from(), t()) :: {:noreply, [], t()}
+  def handle_events(events, _from, state) do
+    # Works in two primary phases. First, we generate fresh t using the
+    # provided events:
+    new_state = update_state(events, state)
+    Logger.debug(["Alert state updated: ", inspect(new_state)])
```

Comment:
Can you make this an `info` actually, and provide a nice search key like `alert_state_updated`? As you mentioned, `server_sent_event_stage` already logs at the debug level, so we could use that potentially if needed, but the app generally doesn't run at debug. Since we only expect a handful of events here a minute, I think, it's fine to just log at info. I suspect we may peek in the logs periodically for this, without wanting to run at full debug logging.

Reply:
Sure.